### PR TITLE
Add SQL operations for new SQL-backed merkle radix implementation

### PR DESCRIPTION
**Pull request by peterschwarz**
This PR adds a set of operations that will support the SQL-backed merkle radix state implementation.  

A future PR will implement the radix tree over a sqlite db using these operations.

### CODE REVIEW

File: libtransact/src/state/merkle/node.rs [hunk 1]
```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2021 Cargill Incorporated
```
Comment (agunde406) [inline libtransact/src/state/merkle/node.rs:2]: Should this include the Intel copyright from the previous file and a date range for Cargill?

File: libtransact/src/state/merkle/sql/migration/sqlite.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        leaf_id INTEGER NOT NULL,
+        from_state_root_id INTEGER NOT NULL,
+        to_state_root_id INTEGER,
+        FOREIGN KEY(from_state_root_id) REFERENCES merkle_radix_state_root(id),
+        FOREIGN KEY(leaf_id) REFERENCES merkle_radix_leaf (id)
+    )
+    "#;
+
+//
+// TODO: Replace with a proper migration solution.
```
Comment (agunde406) [inline libtransact/src/state/merkle/sql/migration/sqlite.rs:58]: Even if it is a TODO, the TODO comment should probably be removed

File: libtransact/src/state/merkle/node.rs (hunk 1 above)
Comment (peterschwarz) [reply libtransact/src/state/merkle/node.rs]: Fixed

File: libtransact/src/state/merkle/sql/migration/sqlite.rs (hunk 2 above)
Comment (peterschwarz) [reply libtransact/src/state/merkle/sql/migration/sqlite.rs]: Fixed

File: libtransact/src/state/merkle/sql/migration/sqlite.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+use diesel::prelude::*;
+
+use crate::error::InternalError;
+
+const CREATE_LEAF_TABLE: &str = r#"
+    CREATE TABLE IF NOT EXISTS merkle_radix_leaf (
+        id INTEGER PRIMARY KEY,
+        address STRING NOT NULL,
+    data BLOB)
```
Comment (vaporos) [inline libtransact/src/state/merkle/sql/migration/sqlite.rs:26]: This looks like it should be indented and the ) on a separate line (when compared to other statements here.

File: libtransact/src/state/merkle/sql/operations/get_path.rs [hunk 4]
```diff
@@ -0,0 +1,290 @@
+/*
```
Comment (vaporos) [inline libtransact/src/state/merkle/sql/operations/get_path.rs:1]: Typo in commit message "Add MerkRadix..." should be "Add MerkleRadix..."

**Review by vaporos — CHANGES_REQUESTED** [review]

File: libtransact/Cargo.toml [hunk 5]
```diff
@@ -32,6 +32,7 @@ bzip2 = { version = "0.4", optional = true }
 cbor-codec = { version = "0.7", optional = true }
 chrono = { version = "0.4", optional = true }
 cylinder = { version = "0.2", optional = true }
+diesel = { version = "1.0", features = ["r2d2"], optional = true }
```
[the remark this answers is not in the document]
Comment (peterschwarz) [reply libtransact/Cargo.toml]: This has been changed to be `1.4.7` explicitly, as that version does not conflict with the rusqlite dependency.

File: libtransact/src/state/merkle/sql/migration/sqlite.rs (hunk 3 above)
Comment (peterschwarz) [reply libtransact/src/state/merkle/sql/migration/sqlite.rs]: Fixed

File: libtransact/src/state/merkle/sql/operations/get_path.rs (hunk 4 above)
Comment (peterschwarz) [reply libtransact/src/state/merkle/sql/operations/get_path.rs:1]: Fixed

File: libtransact/src/state/merkle/sql/migration/sqlite.rs (hunk 3 above)
Comment (agunde406) [reply libtransact/src/state/merkle/sql/migration/sqlite.rs]: looks like the ) still needs to be moved

Comment (peterschwarz) [reply libtransact/src/state/merkle/sql/migration/sqlite.rs]: Fixed

File: libtransact/Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (peterschwarz) [reply libtransact/Cargo.toml]: Added a note before the dependency about this requirement.

**Review by agunde406 — DISMISSED** [review]

File: libtransact/Cargo.toml [hunk 6]
```diff
@@ -51,6 +54,7 @@ sawtooth-sdk = { version = "0.5", optional = true }
 semver = { version = "1", optional = true }
 serde = { version = "1.0", optional = true }
 serde_derive = { version = "1.0", optional = true }
+serde_json = { version = "1.0", optional = true }
```
Comment (vaporos) [inline libtransact/Cargo.toml:57]: This should be "1"

File: libtransact/Cargo.toml [hunk 7]
```diff
@@ -32,6 +32,9 @@ bzip2 = { version = "0.4", optional = true }
 cbor-codec = { version = "0.7", optional = true }
 chrono = { version = "0.4", optional = true }
 cylinder = { version = "0.2", optional = true }
+# diesel 1.4.7 is the minimum version that does not conflict with rusqlite in
+# that they both require a transient dependency of libsqlite-sys.
+diesel = { version = "1.4.7", features = ["r2d2"], optional = true }
```
[the remark this answers is not in the document]
Comment (peterschwarz) [reply libtransact/Cargo.toml]: Fixed

File: libtransact/Cargo.toml (hunk 6 above)
Comment (peterschwarz) [reply libtransact/Cargo.toml]: Fixed

**Review by vaporos — APPROVED** [review]

**Review by agunde406 — APPROVED** [review]

**Pull request merged**